File: # Robert Martin (Uncle Bob)/Clean Architecture p093 (Facade).java
// Subsistema complejo
class PayCalculator {
    public void calculatePay() {
        System.out.println("Calculating pay...");
    }
}

class HourReporter {
    public void reportHours() {
        System.out.println("Reporting hours...");
    }
}

class EmployeeSaver {
    public void saveEmployee() {
        System.out.println("Saving employee data...");
    }
}

// Fachada
class EmployeeFacade {
    private PayCalculator payCalculator;
    private HourReporter hourReporter;
    private EmployeeSaver employeeSaver;

    public EmployeeFacade() {
        payCalculator = new PayCalculator();
        hourReporter = new HourReporter();
        employeeSaver = new EmployeeSaver();
    }

    public void calculatePay() {
        payCalculator.calculatePay();
    }

    public void reportHours() {
        hourReporter.reportHours();
    }

    public void save() {
        employeeSaver.saveEmployee();
    }
}

// Cliente
public class Main {
    public static void main(String[] args) {
        EmployeeFacade facade = new EmployeeFacade();
        facade.calculatePay();
        facade.reportHours();
        facade.save();
    }
}

File: Arquitectura/Robert Martin (Uncle Bob)/Clean Architecture p093 (Facade).cs
using System;

// Subsistema complejo
public class PayCalculator
{
    public void CalculatePay()
    {
        Console.WriteLine("Calculating pay...");
    }
}

public class HourReporter
{
    public void ReportHours()
    {
        Console.WriteLine("Reporting hours...");
    }
}

public class EmployeeSaver
{
    public void SaveEmployee()
    {
        Console.WriteLine("Saving employee data...");
    }
}

// Fachada
public class EmployeeFacade
{
    private PayCalculator payCalculator;
    private HourReporter hourReporter;
    private EmployeeSaver employeeSaver;

    public EmployeeFacade()
    {
        payCalculator = new PayCalculator();
        hourReporter = new HourReporter();
        employeeSaver = new EmployeeSaver();
    }

    public void CalculatePay()
    {
        payCalculator.CalculatePay();
    }

    public void ReportHours()
    {
        hourReporter.ReportHours();
    }

    public void Save()
    {
        employeeSaver.SaveEmployee();
    }
}

// Cliente
class Program
{
    static void Main(string[] args)
    {
        EmployeeFacade facade = new EmployeeFacade();
        facade.CalculatePay();
        facade.ReportHours();
        facade.Save();
    }
}
